main.cc

```diff
@@ -28,106 +29,84 @@
 #include "csv_util.h"
 #include "inifile.h"
 #include "session.h"
+#include "src/core/file.h"
 #include "src/core/usasciicodec.h"
 #include <ctype.h>
 #include <locale.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <signal.h>
+#ifdef AFL_INPUT_FUZZING
+#include "argv-fuzz-inl.h"
+#endif
 
 #define MYNAME "main"
 
 void signal_handler(int sig);
 
-typedef struct arg_stack_s {
-  int argn;
-  int argc;
-  char** argv;
-  struct arg_stack_s* prev;
-} arg_stack_t;
-
-static arg_stack_t
-* push_args(arg_stack_t* stack, const int argn, const int argc, char* argv[])
-{
-  arg_stack_t* res = (arg_stack_t*) xmalloc(sizeof(*res));
-
-  res->prev = stack;
-  res->argn = argn;
-  res->argc = argc;
-  res->argv = (char**)argv;
-
-  return res;
-}
-
-static arg_stack_t
-* pop_args(arg_stack_t* stack, int* argn, int* argc, char** argv[])
+class QargStackElement
 {
-  arg_stack_t* res;
-  char** argv2 = *argv;
-  int i;
+public:
+  int argn;
+  QStringList qargs;
 
-  if (stack == NULL) {
-    fatal("main: Invalid point in time to call 'pop_args'\n");
+public:
+  QargStackElement()
+  {
   }
 
-  for (i = 0; i < *argc; i++) {
-    xfree(argv2[i]);
+  QargStackElement(int p_argn, QStringList p_qargs)
+  {
+    argn = p_argn;
+    qargs = p_qargs;
   }
-  xfree(*argv);
-
-  *argn = stack->argn;
-  *argc = stack->argc;
-  *argv = stack->argv;
+};
 
-  res = stack->prev;
-  xfree(stack);
-
-  return res;
-}
-
-static void
-load_args(const char* filename, int* argc, char** argv[])
+static QStringList
+load_args(const QString& filename, const QString& arg0)
 {
-  gbfile* fin;
-  char* str, *line = NULL;
-  int argc2;
-  char** argv2;
-
-  fin = gbfopen(filename, "r", "main");
-  while ((str = gbfgetstr(fin))) {
-    str = lrtrim(str);
-    if ((*str == '\0') || (*str == '#')) {
+  QString line;
+  QString str;
+  QStringList qargs(arg0);
+
+// Use a QTextStream to read the batch file.
+// By default that will use codecForLocale().
+// By default Automatic Unicode detection is enabled.
+  gpsbabel::File* file = new gpsbabel::File(filename);
+  file->open(QFile::ReadOnly);
+  QTextStream* stream = new QTextStream(file);
+  while (!(str = stream->readLine()).isNull()) {
+    str = str.trimmed();
+    if ((str.isEmpty()) || (str.at(0).toLatin1() == '#')) {
       continue;
     }
-
-    if (line == NULL) {
-      line = xstrdup(str);
+    if (line.isEmpty()) {
+      line = str;
     } else {
-      char* tmp;
-      xasprintf(&tmp, "%s %s", line, str);
-      xfree(line);
-      line = tmp;
+      line.append(' ');
+      line.append(str);
```

Comment:
1. I just literally translated the existing code.
2. The first one is special just to avoid a leading blank.
3. The format of the batch file apparently can have the arguments spread willy nilly over 1 or more linnes.  This code just packs them all into one line so we can parse it.
4. I believe xcsvparse handles quoting, as would be neccesary for a file name containing a blank.  So it is a bit more than split().